ui/article-aem-renderer/src/main/java/org/cru/godtools/article/aem/service/AemArticleManager.kt

```diff
@@ -0,0 +1,316 @@
+
+            // fetch the raw json
+            val t = System.currentTimeMillis().let { if (force) it else it.roundTimestamp(CACHE_BUSTING_INTERVAL_JSON) }
+            val json = try {
+                api.getJson(baseUri.addExtension("9999.json"), t).takeIf { it.code() == HTTP_OK }?.body()
+            } catch (e: IOException) {
+                Timber.tag(TAG).v(e, "Error downloading AEM json")
+                null
+            } ?: return
+
+            // parse & store articles
+            val articles = json.findAemArticles(baseUri).toList()
+            aemDb.aemImportRepository().processAemImportSync(aemImport, articles)
+
+            // launch download of all the articles
+            articles.forEach { coroutineScope.launch { downloadArticle(it.uri, false) } }
+        }
+    }
+
+    init {
+        aemDb.aemImportDao().all
+            .filter { it.isStale() }
+            .forEach { coroutineScope.launch { syncAemImport(it.uri, false) } }
+    }
+    // endregion AEM Import
+
+    // region Download Article
+    @AnyThread
+    suspend fun downloadArticle(uri: Uri, force: Boolean) {
+        articleMutex.withLock(uri) {
+            // short-circuit if there isn't an Article for the specified Uri or if the article doesn't need to be downloaded
+            val article = aemDb.articleDao().find(uri) ?: return
+            if (article.uuid == article.contentUuid && !force) return
+
+            // download the article html
+            try {
+                api.downloadArticle(article.uri.addExtension("html")).takeIf { it.code() == HTTP_OK }?.let { response ->
+                    article.apply {
+                        contentUuid = uuid
+                        content = response.body()
+                        resources = extractResources()
+                    }
+                    aemDb.articleRepository().updateContent(article)
+                    aemDb.resourceDao().getAllForArticle(article.uri).downloadResourcesNeedingUpdate()
+                }
+            } catch (e: IOException) {
+                Timber.tag(TAG).d(e, "Error downloading article")
+            }
+        }
+    }
+    // endregion Download Article
+
+    // region Download Resource
+    @AnyThread
+    protected fun Collection<Resource>.downloadResourcesNeedingUpdate() {
+        filter { it.needsDownload() }.forEach { coroutineScope.launch { downloadResource(it.uri, false) } }
+    }
+
+    @AnyThread
+    suspend fun downloadResource(uri: Uri, force: Boolean) {
+        val resourceDao = aemDb.resourceDao()
+
+        resourceMutex.withLock(uri) {
+            // short-circuit if the resource doesn't exist or it doesn't need to be downloaded
+            val resource = resourceDao.find(uri)
+            if (resource == null || (!force && !resource.needsDownload())) return
+
+            // download the resource
+            withContext(Dispatchers.IO) {
```

Comment:
If I remember correctly, the IO dispatcher is specifically used for network requests, right?

Reply:
yeah, or reading data from the disk. basically any "blocking" operation should use `Dispatchers.IO`